File: message_consumer_node/src/service/consumer.ts
import amqb from 'amqplib/callback_api'
import { db } from '../db/connection';
import { RequestLogs } from '../db/schema';
import { app } from '..';

const amqbHost = process.env.AMQBHOST || ''
const QUEUE_NAME =  process.env.QUEUE_NAME ||''

let connection : amqb.Connection | null
let logsBuffer: Array<any> = [];
const BULK_INSERT_THRESHOLD = 10;
let isInserting = false;

export default function startConsuming(){
    amqb.connect(amqbHost, (err , conn) => {
        if(err)throw err
        connection = conn
        conn.createChannel((err, channel) => {
            if(err)throw err

            channel.assertQueue(QUEUE_NAME, {
                durable: true
            })

            console.log(" [*] Waiting for messages in %s. To exit press CTRL+C", QUEUE_NAME);

            channel.consume(QUEUE_NAME, async (msg) => {
                if(!msg) return;

                const logData = JSON.parse(msg?.content.toString())

                logsBuffer.push({
                    method: logData.method,
                    path: logData.path,
                    headers: JSON.stringify(logData.headers),
                    body: logData.body,
                    response_status: logData.response_status,
                    response_body: logData.response_body,
                    timestamp: logData.timestamp,
                })

                if(logsBuffer.length == BULK_INSERT_THRESHOLD && !isInserting){
                    isInserting = true;
                    await save_logs_bulk();
                }
                    channel.ack(msg)
              }, {
                  noAck: false
              });
        })

    })
}


async function save_logs_bulk() {
    if(logsBuffer.length == 0) return;
    const logsToInsert = logsBuffer.slice(0,BULK_INSERT_THRESHOLD)
    try {
        await db.insert(RequestLogs).values(logsToInsert)
        console.log(`✅ Inserted ${logsBuffer.length} logs into DB`);
        logsBuffer = logsBuffer.slice(BULK_INSERT_THRESHOLD)
    } catch (error) {
        console.error("❌ Error inserting logs:", error);
    } finally{
        isInserting = false;
        if(logsBuffer.length > BULK_INSERT_THRESHOLD){
            await save_logs_bulk()
        }
    }
}

async function shutdown() {
    console.log('Shutting down...');
    if(logsBuffer.length > 0){
        console.log(`Inserting remaining ${logsBuffer.length} logs into DB`);
        await save_logs_bulk();
    }
    if(connection){
        connection.close();
    }
    app.close()
    process.exit(0);
}
process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);
process.on('uncaughtException', (err) => {
    console.error('Uncaught exception:', err);
    shutdown();
});

File: message_producer_django/core/logger/RabbitMQ/consumer.py
import pika
import json

import pika.exceptions
from ..models import RequestLog
from django.db import transaction
import threading
import sys
import signal
import time

RABBITMQ_HOST = 'localhost'
QUEUE_NAME = 'request_logs'

logs = []
logs_lock = threading.Lock()
connection = None
channel = None

def save_logs_bulk():
    global logs
    with logs_lock:
        if logs:
            bulk_logs = [
                RequestLog(
                    method=log["method"],
                    path=log["path"],
                    headers=json.dumps(log["headers"]),
                    body=log["body"],
                    response_status=log["response_status"],
                    response_body=log["response_body"],
                    timestamp=log["timestamp"]
                ) for log in logs
            ]

            with transaction.atomic():
                RequestLog.objects.bulk_create(bulk_logs)

            print(f"✅ Inserted {len(logs)} logs into DB")
            logs.clear()


def consume_message():
    while True:
        try:
            connection = pika.BlockingConnection(pika.ConnectionParameters(host=RABBITMQ_HOST))
            channel = connection.channel()
            channel.queue_declare(queue=QUEUE_NAME, durable=True)
            print("is consumer working")
            logs=[]

            def callback(ch, method, properties, body):
                log_data = json.loads(body)
                with logs_lock:
                    logs.append(log_data)

                if len(logs) >= 10:
                    save_logs_bulk()
                    logs.clear()
                    print("✅ Bulk inserted 10 logs into DB")
        
                ch.basic_ack(delivery_tag=method.delivery_tag)

            channel.basic_consume(queue=QUEUE_NAME,on_message_callback=callback)
            channel.start_consuming()
        except pika.exceptions.AMQPConnectionError as e:
            print(f"Error connecting to RabbitMQ: {e}")
            time.sleep(5)
        except Exception as e:
            print(f"❌ Unexpected error: {e}")
            break



def graceful_shutdown(signal_received, frame):
    print("\n🚦 Shutdown signal received. Saving remaining logs...")
    save_logs_bulk()
    if connection:
        connection.close()
    sys.exit(0)

signal.signal(signal.SIGINT, graceful_shutdown)
signal.signal(signal.SIGTERM, graceful_shutdown)

def start_consumer_thread():
    consumer_thread = threading.Thread(target=consume_message)
    consumer_thread.daemon=True
    consumer_thread.start()

